Quiz App/src/pages/MainQuiz/MainQuiz.tsx

```diff
@@ -1,108 +1,198 @@
     default:
       return state;
   }
 };
 
+interface MainQuizProps {}
+
 const MainQuiz: React.FC<MainQuizProps> = () => {
   const [state, dispatch] = useReducer(reducer, initialState);
+  const { progress, numberOfQuestions, currentQuestionId, isModalOpen } = state;
+  const { startTimer, stopTimer, resetTimer, getElapsedTime } = useTimer();
+  const config = useAppSelector((state) => state.quizConfig.configuration);
+  const { data, error, isLoading } = quizApi.useFetchQuestionsQuery(
+    config as Configuration,
+  );
+  const timeLimit = config!.time;
 
-  const timeLimit = 140;
-  const currentQuestion = quizQuestion[state.currentQuestionId];
+  const dispatchAnswers = useAppDispatch();
+  const navigate = useNavigate();
 
-  const selectAnswer = () => {
-    state.currentQuestionId < quizQuestion.length - 1
-      ? dispatch({
-          type: "CURRENT_QUESTION_ID",
-          payload: { currentQuestionId: state.currentQuestionId + 1 },
-        })
-      : endQuiz();
-  };
-  const endQuiz = () => {};
+  const currentQuestionData = useMemo(() => {
+    return data ? data[currentQuestionId] : null;
+  }, [data, currentQuestionId]);
+
+  useEffect(() => {
+    if (data) {
+      dispatchAnswers(setTotal(data.length));
+      dispatch({ type: "SET_NUMBER_OF_QUESTIONS", payload: data.length });
+      resetTimer();
+      startTimer();
+    }
+  }, [data]);
 
   useEffect(() => {
-    const uniqueIds = new Set(quizQuestion.map((question) => question.id));
+    dispatch({ type: "SET_PROGRESS", payload: currentQuestionId + 1 });
+  }, [currentQuestionId]);
 
-    dispatch({
-      type: "QUIZ__STATE",
-      payload: {
-        numberOfQuestions: uniqueIds.size,
-        progress: state.currentQuestionId + 1,
-      },
-    });
-  }, [state.currentQuestionId]);
+  const endQuiz = () => {
+    dispatch({ type: "TOGGLE_MODAL", payload: true });
+  };
+
+  const confirm = () => {
+    dispatchAnswers(resetQuiz());
+    dispatchAnswers(clearQuizData());
+    navigate("/");
+  };
+
+  const cancel = () => {
+    dispatch({ type: "TOGGLE_MODAL", payload: false });
+  };
+
+  const timeOrQuestionsEnd = useCallback(() => {
+    stopTimer();
+    dispatchAnswers(setTime(getElapsedTime()));
+    resetTimer();
+    navigate("/result");
+  }, [dispatchAnswers]);
+
+  const selectAnswer = useCallback(
+    (answer: string) => {
+      if (currentQuestionData) {
+        const isCorrect = currentQuestionData.correctAnswer === answer;
+
+        if (isCorrect) {
+          dispatchAnswers(setCorrectAnswers());
+        }
+
+        if (currentQuestionId < numberOfQuestions - 1) {
+          dispatch({
+            type: "SET_CURRENT_QUESTION_ID",
+            payload: currentQuestionId + 1,
+          });
+        } else {
+          timeOrQuestionsEnd();
+        }
+      }
+    },
+    [
+      currentQuestionData,
+      currentQuestionId,
+      numberOfQuestions,
+      dispatchAnswers,
+      timeOrQuestionsEnd,
+    ],
+  );
```

Comment:
```suggestion
const setNextQuestionOrEnd = useCallback(() => {
  currentQuestionId < numberOfQuestions - 1
  ? dispatch({ type: "SET_CURRENT_QUESTION_ID", payload: currentQuestionId + 1 })
  : timeOrQuestionsEnd();
}, [currentQuestionId, numberOfQuestions, timeOrQuestionsEnd, dispatch]);

const checkAnswer = useCallback((answer: string) => {
  if (!currentQuestionData) return;
  
  const isCorrect = currentQuestionData.correctAnswer === answer;
  if (isCorrect) dispatchAnswers(setCorrectAnswers());
}, [currentQuestionData, dispatchAnswers]);

const selectAnswer = useCallback((answer: string) => {
  checkAnswer(answer);
  setNextQuestionOrEnd();
}, [checkAnswer, setNextQuestionOrEnd]);
```
wdyt about splitting this one into 3 new?